Major artifact rule update to align with specs

lukpueh opened the pull request:
Aligns with latest specs and discussion in https://github.com/in-toto/in-toto/issues/43

**Changes in verifylib**
- Artifact rules updates
   - Changes MATCH rule format and behavior 
   ```
   MATCH <pattern> [IN <source-path-prefix>] WITH (MATERIALS|PRODUCTS) 
    [IN <destination-path-prefix>] FROM <step>
    ```
  - Changes CREATE/DELETE rule behavior
  - Re-adds MODIFY rule
  - Adds ALLOW/DISALLOW rules
- Adds hashdict equality check function
- Changes verify_item_rules arguments
  Receives additional source_type argument to know if the rules were in material_matchrules or 
  product_matchrules of a step or inspection.
  Removes artifacts arguments. Source artifacts are now taken from the passed links because they 
  contains all links now (steps and inspections).
- Changes verify_item_rules algorithm
  Does not return when artifact queue is empty. This prevents rules from passing just because they
  were skipped.
- Changes verify_all_item_rules arguments
  Receives only one link dictionary for step links and inspection links.

**Various related changes**
- Replaces matchrule_validators.py with artifact_rules.py
   Its function unpack_rule is now used instead of check_matchrule_syntax
  to parse all rules for rule data and check syntax.
 - Removes flatten and invert utility function. 
    Matching solely by hash leads to collisions.
 - Updates demo to use new rules
 - Updates unit tests

Commit:
Major artifact rule update to align with specs

* Various changes in verifylib:
- Artifact rules:
  Changes MATCH rule format and behavior
    MATCH <pattern> [IN <source-path-prefix>] WITH (MATERIALS|PRODUCTS)
    [IN <destination-path-prefix>] FROM <step>
  Changes CREATE/DELETE rule behavior
  Re-adds MODIFY rule
  Adds ALLOW/DISALLOW rules

- Changes verify_item_rules arguments
  Receives additional source_type argument to know if the rules were in
  material_matchrules or product_matchrules of a step or inspection.
  Removes artifacts arguments. Source artifacts are now taken from the
  passed links because they contains all links now (steps and inspections).

- Changes verify_item_rules algorithm
  Does not return when artifact queue is empty. This prevents rules from
  passing just because they were skipped.

- Changes verify_all_item_rules arguments
  Receives only one link dictionary for step links and inspection links.

* Replaces matchrule_validators.py with artifact_rules.py
  Its function unpack_rule is now used instead of check_matchrule_syntax
  to parse all rules for rule data and check syntax.

* Removes flatten and invert utility function
  Matching solely by hash leads to collisions.

* Updates demo create_layout.py to use new rules

* Updates unit tests

SantiagoTorres commented on demo/owner_alice/create_layout.py:
```diff
@@ -36,21 +36,20 @@ def main():
             # files. But we do have to think of a way to ignore
             # irrelevant files.
-            ["CREATE", "*"],
+            ["ALLOW", "*"],
```
do we still need this?

SantiagoTorres commented on demo/owner_alice/create_layout.py:
```diff
@@ -36,21 +36,20 @@ def main():
+            ["MATCH", "foo.py", "WITH", "PRODUCTS", "FROM", "write-code"],
             # FIXME: See material_matchrules above
-            ["CREATE", "*"],
+            ["ALLOW", "*"],
```
also here

SantiagoTorres commented on in_toto/artifact_rules.py:
```diff
@@ -0,0 +1,186 @@
+  in_toto/artifact_rules.py
+
+<Author>
+  Santiago Torres-Arias <santiago@nyu.edu>
```
considering all the refactoring done, maybe it would be good to add your name here

SantiagoTorres commented on in_toto/verifylib.py:
```diff
@@ -306,294 +308,582 @@ def verify_all_steps_command_alignment(layout, links_dict):
-    general understanding of glob patterns (especially "*").
+def _equal_hash_dicts(hash_dict_a, hash_dict_b):
+  """Compare to hash_dict for equality, formats are
+  securesystemslib.formats.HASHDICT_SCHEMA. Return True or False. """
```
Is this check being made before? or should we run this check before doing anything else.

lukpueh replied:
What check are you referring to? `securesystemslib.formats. HASHDICT_SCHEMA.check_match`?

SantiagoTorres commented on in_toto/verifylib.py:
```diff
@@ -306,294 +308,582 @@ def verify_all_steps_command_alignment(layout, links_dict):
-    source artifact with a matching hash in the passed artifact dictionary and
-    that this artifact in also the artifact queue.
+  for key in hash_dict_a.keys():
+    if hash_dict_a[key] != hash_dict_b[key]:
```
I assume there are no recursive properties, but could we short_circuit to check for equality in both the value and the reference? 

I'm also not sure if this method is [strictly necessary](https://github.com/python/cpython/blob/master/Objects/dictobject.c#L2697-L2737)

lukpueh replied:
I had this discussion with @awwad the other day. For some reason I was convinced that dictionaries had to be compared manually, apparently I was wrong.
I'll remove the function altogether.

SantiagoTorres commented on in_toto/verifylib.py:
```diff
@@ -306,294 +308,582 @@ def verify_all_steps_command_alignment(layout, links_dict):
+          " in materials but should have been newly created."
+          .format(rule, matched_product))
+
+  return list(set(source_products_queue) - set(matched_products))
```
I wonder if this would be easier to do using pop() (maybe deepcopying the list). This may just be a matter of style though.

lukpueh replied:
I'm not married to having the set difference operation, especially as we need to temporarily cast. Using pip would require an additional loop over the list, though. I think we can leave it for now.

SantiagoTorres replied:
The issue is that this is not a cast, you are allocating a new set object for both lists, and then turning the resulting set object into another list. Of course this is a minor detail, but it sounds expensive/heavy.

SantiagoTorres reviewed (changes requested):
I have minor comments. Given that this PR is so big, I want to make another pass once these are addressed....

SantiagoTorres commented on in_toto/artifact_rules.py:
```diff
@@ -0,0 +1,186 @@
+
+  if not isinstance(pattern, six.string_types):
+    raise securesystemslib.exceptions.FormatError("Wrong rule format,"
+        " 'pattern' (second element) must be a string.\n"
```
I wonder if here we want to say "the pattern". It feels oddly worded here.

SantiagoTorres commented on in_toto/artifact_rules.py:
```diff
@@ -0,0 +1,186 @@
+
+    # ... IN <source-path-prefix> WITH (MATERIALS|PRODUCTS)
+    # IN <destination-path-prefix> FROM <step>
+    if (rule_len == 10 and rule[2].lower() == "in" and
```
I wonder if it would improve readabilty to lower() all the strings before starting comparisons. e.g.,
```
rule = [x.lower for x in rule]
```

lukpueh replied:
List comprehension!!

lukpueh replied:
I thought about this too, but I think we shouldn't convert the user defined parts (pattern, prefix, step).

SantiagoTorres replied:
we are making a copy of it here. Or what do you mean?

lukpueh replied:
Making a copy of what where?

I meant two things:
1. List comprehension is discouraged by our code  style guidelines
2. I think the user defined parts of the rule "path pattern", "path prefix" and "step name" should not be converted to lower, because the case might matter for those.
Hence we'd have to explicitly convert certain elements that are at varying indices depending on the rule type or whether it has optional parts in it or not.

SantiagoTorres commented on in_toto/artifact_rules.py:
```diff
@@ -0,0 +1,186 @@
+          " (MATERIALS|PRODUCTS) [IN <destination-path-prefix>] FROM <step>.\n"
+          "Got: \n\t{}".format(rule))
+
+    if not isinstance(dest_type, six.string_types) or (
```
wouldn't it be nicer to iterate over all the rules to check they are strings first?

SantiagoTorres commented on in_toto/models/layout.py:
```diff
@@ -93,6 +93,7 @@ def dump(self, filename='root.layout'):
   def read_from_file(filename='root.layout'):
     """Static method to instantiate a new Layout object from a
     canonical JSON serialized file """
+    print filename
```
I don't believe we want this.

SantiagoTorres commented on test/models/test_artifact_rules.py:
```diff
@@ -0,0 +1,191 @@
+  Test artifact rule unpacking.
+
+"""
+
```
I see this blank space in some files but not in others. Does it matter?

lukpueh replied:
Yep, we [want this](https://github.com/secure-systems-lab/code-style-guidelines#documentation-strings). But I won't change this everywhere in this PR. This could be candidate for a general code cleanup PR.
> Note that most importantly, the """ that ends a multiline docstring should be on a line by itself, and preferably preceded by a blank line, e.g.:

SantiagoTorres replied:
It says preceded, I'm talking about the succedding line before the imports.

lukpueh replied:
Ah ok. I guess I can remove that. :)

SantiagoTorres commented on test/models/test_artifact_rules.py:
```diff
@@ -0,0 +1,191 @@
+
+  def test_unpack_rule_unknown_rule_type(self):
+    """Test generic rule syntax error, too many arguments. """
+    rule = ["SUBVERT", "foo"]
```
This should be a valid rule tbh

lukpueh replied:
?

SantiagoTorres replied:
ignore this, it's a silly comment 🚀

SantiagoTorres commented on test/test_verifylib.py:
```diff
@@ -156,121 +159,275 @@ def test_commands_do_not_align_at_all_log_warning(self):
+    sha_1 = ("d65165279105ca6773180500688df4bd"
+                  "c69a2c7b771752f0a46ef120b7fd8ec3")
 
+    sha_2 = ("155c693a6b7481f48626ebfc545f0523"
```
do we want to change these names to something that doesn't hint of either algorithm?

SantiagoTorres commented on test/test_verifylib.py:
```diff
@@ -307,231 +469,280 @@ def setUp(self):
+    self.sha256_1 = \
         "d65165279105ca6773180500688df4bdc69a2c7b771752f0a46ef120b7fd8ec3"
-    self.sha256_bar = \
+    self.sha256_2 = \
```
Like this, I think this name is better.

Commit:
Adapts demo layout creation script to new rules

- Updates RULE syntax for match rules
- Replaces CREATE rule with new ALLOW rule where applicable
- Adds explicit ALLOW rules in inspection

Commit:
Adds Lukas as author in artifact_rules module

Commit:
Removes _equal_hash_dicts internal function

_equal_hash_dicts was used to manually compare hash dictionaries,
but we might as well use plain Python dict comparison.
github.com/python/cpython/blob/master/Objects/dictobject.c#L2697-L2737

Commit:
Slightly modifies artifact rule exception messages

Adds determiners to nouns.

Commit:
Removes debug print statement in layout

Commit:
Merges rule element string checks

Commit:
Re-adds hash algo type for unit test variables

It's a sha256, so let's call it a sha256.

lukpueh commented:
@SantiagoTorres thanks for your comments. Ready for another round?

Commit:
Removes blank lines before import block

SantiagoTorres commented on in_toto/artifact_rules.py:
```diff
@@ -0,0 +1,171 @@
+  #else:
+  # can't reach else
+  # if it is neither in GENERIC_RULES nor in MATCH_RULES it would
+  # have already raised an exception above when checking if it is in ALL_RULES
```
I guess this else is left for the sake of documentation, but do we really want it?

Commit:
Adds rule lowercase copy for easier parsing

Create all lower rule copy to case insensitively parse out tokens
whose position we don't know at that time.
We keep the original rule to retain the non-token elements' case.
